backend/internal/authn/assert/generator.go

```diff
@@ -0,0 +1,201 @@
+func (ag *authAssertGenerator) VerifyAssurance(context *AssuranceContext, requiredAAL AssuranceLevel,
+	requiredIAL AssuranceLevel) (bool, *serviceerror.ServiceError) {
+	logger := log.GetLogger().With(log.String(log.LoggerKeyComponentName, loggerComponentName))
+	logger.Debug("Verifying assurance levels")
+
+	if context == nil {
+		logger.Debug("Nil assurance context provided")
+		return false, &ErrorNilAssuranceContext
+	}
+	if requiredAAL == "" && requiredIAL == "" {
+		logger.Debug("No assurance levels specified for verification")
+		return false, &ErrorNoAssuranceRequirements
+	}
+
+	// Check AAL level
+	if requiredAAL != "" && !ag.meetsAssuranceLevel(context.AAL, requiredAAL) {
+		logger.Debug("Actual AAL does not meet required AAL", log.String("actualAAL", string(context.AAL)),
+			log.String("requiredAAL", string(requiredAAL)))
+		return false, nil
+	}
+
+	// Check IAL level
+	if requiredIAL != "" && !ag.meetsAssuranceLevel(context.IAL, requiredIAL) {
+		logger.Debug("Actual IAL does not meet required IAL", log.String("actualIAL", string(context.IAL)),
+			log.String("requiredIAL", string(requiredIAL)))
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// extractUniqueAuthenticators extracts unique authenticators and factors from authenticator references.
+// returns a map of unique authenticator names and a set of unique authentication factors.
+func (ag *authAssertGenerator) extractUniqueAuthenticators(authenticators []authncm.AuthenticatorReference,
+	logger *log.Logger) (map[string]bool, map[authncm.AuthenticationFactor]bool) {
+	authenticatorsMap := make(map[string]bool)
+	factorSet := make(map[authncm.AuthenticationFactor]bool)
+
+	for _, auth := range authenticators {
+		authenticatorsMap[auth.Authenticator] = true
+
+		factors := authncm.GetAuthenticatorFactors(auth.Authenticator)
+		if len(factors) == 0 {
+			logger.Debug("No factors found for authenticator. Skipping",
+				log.String("authenticator", auth.Authenticator))
+			continue
+		}
+
+		for _, factor := range factors {
+			factorSet[factor] = true
+		}
+	}
+
+	return authenticatorsMap, factorSet
+}
+
+// calculateAAL calculates the AAL based on the authentication factors.
+// - UNKNOWN: No valid authentication factors found
+// - AAL1: Single-factor authentication (any one factor)
+// - AAL2: Two-factor authentication (two different factors)
+// - AAL3: Multi-factor authentication with hardware-based cryptographic authenticator
+func (ag *authAssertGenerator) calculateAAL(factorSet map[authncm.AuthenticationFactor]bool,
+	logger *log.Logger) AssuranceLevel {
+	var aal AssuranceLevel
+	factorCount := len(factorSet)
+
+	switch factorCount {
+	case 0:
+		aal = AALUnknown
+	case 1:
+		aal = AALLevel1
+	case 2:
+		aal = AALLevel2
+	default:
+		aal = AALLevel3
+	}
+
+	if logger.IsDebugEnabled() {
+		factorNames := make([]string, 0, len(factorSet))
+		for factor := range factorSet {
+			factorNames = append(factorNames, string(factor))
+		}
```

Comment:
No need to convert this to array, if `extractUniqueAuthenticators()` returned an array in the first place.